File: donghwa/qual/2106/chatgpt/01_smallest_int.py
def find_smallest_positive(arr):
    n = len(arr)
    
    # Separate positive and non-positive numbers
    pos_nums = [x for x in arr if x > 0]
    
    # Create a boolean array to mark the presence of positive numbers
    present = [False] * (n + 1)
    
    # Mark the presence of positive numbers in the array
    for num in pos_nums:
        if 1 <= num <= n:
            present[num] = True

    # Find the smallest positive integer that is not in the array
    for i in range(1, n + 2):
        if not present[i]:
            return i

    return 1

# Example usage:
arr1 = [4, 7, -1, 1, 2]
print("Example 1:", find_smallest_positive(arr1))

arr2 = [100, 101, 102]
print("Example 2:", find_smallest_positive(arr2))

arr3 = [3, 2, 1, 0, -1]
print("Example 3:", find_smallest_positive(arr3))

File: donghwa/qual/2106/chatgpt/01_smallest_int.cpp
#include <stdio.h>
#include <iostream>
#include <vector>

using namespace std;

int segregate(vector<int>& nums) {
    int n = nums.size();
    for (int i = 0; i < n; i++) {
        while (nums[i] > 0 && nums[i] <= n && nums[nums[i] - 1] != nums[i]) {
            swap(nums[i], nums[nums[i] - 1]);
        }
    }
    for (int i = 0; i < n; i++) {
        if (nums[i] != i + 1) {
            return i + 1;
        }
    }
    return n + 1;
}

int findSmallestPositive(vector<int>& nums) {
    int n = nums.size();
    if (n == 0) {
        return 1;
    }
    
    int shift = segregate(nums);
    vector<int> positive_nums(nums.begin() + shift, nums.end());

    return segregate(positive_nums);
}

int main() {
    // Example 1
    vector<int> arr1 = {4, 7, -1, 1, 2};
    cout << "Example 1: " << findSmallestPositive(arr1) << endl;

    // Example 2
    vector<int> arr2 = {100, 101, 102};
    cout << "Example 2: " << findSmallestPositive(arr2) << endl;

    // Example 3
    vector<int> arr3 = {3, 2, 1, 0, -1};
    cout << "Example 3: " << findSmallestPositive(arr3) << endl;

    return 0;
}
